include/geos/util/Vector.h

```diff
@@ -0,0 +1,256 @@
+        }
+        m_size += static_cast<std::uint32_t>(n);
+
+        return pos;
+    }
+
+    void insert(T* pos, std::size_t n, const T& value) {
+        pos = make_space(pos, n);
+        std::fill(pos, pos + n, value);
+    }
+
+    template<typename Iter>
+    void insert(T* pos, Iter from, Iter to) {
+        auto n = static_cast<std::size_t>(to - from);
+
+        if (from >= begin() && from < end()) {
+            // from and to may be invalidated
+            auto from_n = from - begin();
+            auto to_n = to - begin();
+
+            pos = make_space(pos, n);
+
+            from = begin() + from_n;
+            to = begin() + to_n;
+        } else {
+            pos = make_space(pos, n);
+        }
+
+        std::copy(from, to, pos);
+    }
+
+    template<typename Iter>
+    void assign(Iter from, Iter to) {
+        assert(static_cast<std::size_t>(to - from) <= size());
+        std::copy(from, to, begin());
+    }
+
+    void reserve(std::size_t sz) {
+        if (sz <= capacity()) {
+            return;
+        }
+
+        T* tmp = sz > 0 ? new T[sz] : nullptr;
+        if (tmp && !empty()) {
+            std::memcpy(tmp, m_buf, m_size * sizeof(T));
+        }
+        if (owned()) {
+            delete[] m_buf;
+        }
+        m_buf = tmp;
+        m_capacity = static_cast<std::uint32_t>(sz);
+    }
+
+    void resize(std::size_t sz) {
+        reserve(sz);
+        m_size = static_cast<std::uint32_t>(sz);
+    }
+
+    void clear() {
+        m_size = 0;
+    }
+
+    std::size_t capacity() const {
+        return m_capacity;
+    }
+
+    std::size_t size() const {
+        return m_size;
+    }
+
+    bool empty() const {
+        return m_size == 0;
+    }
+
+    bool owned() const {
+        return data() == nullptr || !(capacity() == 0 && size() > 0);
+    }
+
+    const T& operator[](std::size_t i) const {
+        return *(data() + i);
+    }
+
+    T& operator[](std::size_t i) {
+        return *(data() + i);
+    }
+
+    T* release() {
+        m_capacity = 0;
+        return m_buf;
+    }
+
+    T* data() {
+        return m_buf;
+    }
+
+    const T* data() const {
+        return m_buf;
+    }
+
+    T* begin() {
+        return data();
+    };
+
+    T* end() {
+        return data() + size();
+    }
+
+    const T* begin() const {
+        return data();
+    };
+
+    const T* end() const {
+        return data() + size();
+    }
+
+private:
+
+    void growIfNeeded(std::size_t num_to_add) {
+        if (size() + num_to_add > capacity()) {
+            auto new_capacity = capacity() == 0 ?
+                        std::max(size() + num_to_add, static_cast<std::size_t>(4)) :
+                        static_cast<std::size_t>(static_cast<double>(capacity()) * 1.5);
+            new_capacity = std::max(new_capacity, capacity() + num_to_add);
+            reserve(new_capacity);
+        }
+    }
+
+    T* m_buf;
+    std::uint32_t m_capacity;
```

Comment:
why not use size_t for m_capacity and m_size ?

Reply:
The C API uses 32 bits for sizes and incides, I'm doubtful a coordinate sequence with > 1 billion points would function in practice, and I'm trying to be sensitive to the object size of `Point`, whose API requires that it be backed by a `CoordinateSequence`.